api/envoy/config/route/v3/route_components.proto

```diff
@@ -2228,3 +2229,31 @@ message FilterConfig {
   // than rejecting the config.
   bool is_optional = 2;
 }
+
+// The policy that used to determine which name or value will be used as the searching key to get per filter config
+// from :ref:`VirtualHost.typed_per_filter_config<envoy_v3_api_field_config.route.v3.VirtualHost.typed_per_filter_config>`
+// and :ref:`Route.typed_per_filter_config<envoy_v3_api_field_config.route.v3.Route.typed_per_filter_config>`
+// at runtime. The ``fallback_policy`` will be used by default.
+// [#not-implemented-hide:]
+message FilterConfigKeyPolicy {
+  oneof key_policy {
+    // Both :ref:`custom filter config names
+    // <envoy_v3_api_field_extensions.filters.network.http_connection_manager.v3.HttpFilter.name>` and canonical
+    // filter names (e.g., envoy.filters.http.router) will be used. The custom filter config names will be used
+    // first. And if no per filter config has been found via these custom names then the canonical filter names
+    // will be used as fallback.
+    google.protobuf.Empty fallback_policy = 1;
+    // Canonical filter names (e.g., envoy.filters.http.router) will be used as the searching key of per filter
+    // config.
+    google.protobuf.Empty filter_name_key = 2;
+    // :ref:`custom filter config names
+    // <envoy_v3_api_field_extensions.filters.network.http_connection_manager.v3.HttpFilter.name>` will be used
+    // as the searching key of per filter config.
+    google.protobuf.Empty custom_name_key = 3;
+
+    // Extendable custom provider that will used to generate a searching key from custom filter config names,
+    // canonical filter names, or even request headers.
+    // [#not-implemented-hide:]
+    core.v3.TypedExtensionConfig custom_provider = 4;
```

Comment:
Do we have a use-case for this?

Also If both `filter_name_key ` and `fallback_policy`'s primary goal is backward compatibility, then I didn't see the value of them except for adding more complexity for the user. Then I would prefer to use the runtime key or only add the fallback logic, then wait for the end-user to move to the final target in the future.

Reply:
> Do we have a use-case for this?

A simple scene is per user/tenant route level config. We can get user/tenant info  after authorization/authentication, then select a route level fitler config based on the filter name and user/tenant info.

Reply:
another solution is we use the fallback logic, and disable searching by name in the future. And for `custom_provider`, it can be a separate story. Just not sure `filter_name_key` and `custom_name_key` is worth or not.

Reply:
I think it would be very hard to disable serching by filter name completely. Too much users will be effected.

Reply:
What I think is that this field can be ignored by most of normal users. It will not bring any new complexity to these users.

And for some proficient users, this field can provides complete control to the searching behaviour.